File: src/main/java/com/intgrah/algorithms/list/LinkedList.java
package com.intgrah.algorithms.list;

import java.util.Iterator;

public class LinkedList<T> extends AbstractList<T> implements Stack<T, Void>, Queue<T, Void> {

    private Node front;
    private Node back;

    public T getFront() {
        assert !isEmpty();
        return front.value;
    }

    public Void pushFront(T value) {
        Node n = new Node(value);
        n.next = front;
        front = n;
        if (size++ == 0)
            back = n;
        return null;
    }

    public T popFront() {
        assert !isEmpty();
        Node f = front;
        front = f.next;
        if (--size == 0)
            back = null;
        return f.value;
    }

    public T getBack() {
        assert !isEmpty();
        return back.value;
    }

    public Void pushBack(T value) {
        Node n = new Node(value);
        if (size++ == 0)
            front = n;
        else
            back.next = n;
        back = n;
        return null;
    }

    public void append(LinkedList<T> l) {
        if (size == 0)
            front = l.front;
        else
            back.next = l.front;
        back = l.back;
        size += l.size;
        l.clear();
    }

    public void clear() {
        front = back = null;
        size = 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private Node n = front;

            @Override
            public boolean hasNext() { return n != null; }

            @Override
            public T next() {
                T value = n.value;
                n = n.next;
                return value;
            }
        };
    }

    private class Node {

        private final T value;
        private Node next;

        Node(T v) { value = v; }

    }

}

File: list/linked_list.py
from collections.abc import Iterator


class SingleNode[T]:
    def __init__(self, value: T) -> None:
        self.value = value
        self.next: SingleNode[T] | None = None


class LinkedList[T]:
    def __init__(self) -> None:
        self.front: SingleNode[T] | None = None
        self.len = 0

    def push_front(self, value: T) -> SingleNode[T]:
        front = SingleNode(value)
        front.next = self.front
        self.front = front
        self.len += 1
        return front

    def pop_front(self) -> T:
        if self.front is None:
            raise IndexError
        value = self.front.value
        self.front = self.front.next
        self.len -= 1
        return value
    
    def clear(self) -> None:
        self.front = None
        self.len = 0
    
    def __iter__(self) -> Iterator[T]:
        n = self.front
        while n is not None:
            yield n.value
            n = n.next
            
    def __len__(self) -> int:
        return self.len
